File: backend/app/core/permissions.py
"""平台角色与权限定义"""

from app.models.user import User, UserRole

# 角色显示名（API / 文档参考）
ROLE_LABELS: dict[UserRole, str] = {
    UserRole.SUPER_ADMIN: "超级管理员",
    UserRole.ADMIN: "管理员",
    UserRole.MANAGER: "项目经理",
    UserRole.ANNOTATOR: "标注员",
    UserRole.REVIEWER: "审核员",
}

# 各角色可访问的平台能力
PLATFORM_PERMISSIONS: dict[str, set[str]] = {
    UserRole.SUPER_ADMIN.value: {
        "users.manage",
        "users.assign_role",
        "users.reset_password",
        "projects.manage",
        "tasks.manage",
        "export.run",
        "quality.run",
    },
    UserRole.ADMIN.value: {
        "users.manage",
        "users.assign_role",
        "users.reset_password",
        "projects.manage",
        "tasks.manage",
        "export.run",
        "quality.run",
    },
    UserRole.MANAGER.value: {
        "projects.manage",
        "tasks.manage",
        "export.run",
    },
    UserRole.REVIEWER.value: {
        "tasks.review",
    },
    UserRole.ANNOTATOR.value: set(),
}

# 管理员可分配的角色（不含 super_admin）
ADMIN_ASSIGNABLE_ROLES = {
    UserRole.MANAGER,
    UserRole.ANNOTATOR,
    UserRole.REVIEWER,
    UserRole.ADMIN,
}

SUPER_ADMIN_ONLY_ROLES = {UserRole.SUPER_ADMIN}


def role_value(role: UserRole | str) -> str:
    return role.value if isinstance(role, UserRole) else role


def has_platform_permission(user: User, permission: str) -> bool:
    perms = PLATFORM_PERMISSIONS.get(role_value(user.role), set())
    return permission in perms


def can_assign_role(actor: User, target_role: UserRole) -> bool:
    """判断操作者是否可将用户设为 target_role。"""
    if actor.role == UserRole.SUPER_ADMIN:
        return True
    if actor.role == UserRole.ADMIN:
        return target_role in ADMIN_ASSIGNABLE_ROLES
    return False

File: frontend/src/utils/permissions.ts
/** 平台角色与权限（与后端 UserRole / PLATFORM_PERMISSIONS 对齐） */

export type PlatformRole =
  | 'super_admin'
  | 'admin'
  | 'manager'
  | 'annotator'
  | 'reviewer'

export const ROLE_LABELS: Record<PlatformRole, string> = {
  super_admin: '超级管理员',
  admin: '管理员',
  manager: '项目经理',
  annotator: '标注员',
  reviewer: '审核员',
}

export const ROLE_COLORS: Record<PlatformRole, string> = {
  super_admin: '#f59e0b',
  admin: '#ef4444',
  manager: '#00d4ff',
  annotator: '#10b981',
  reviewer: '#a78bfa',
}

export const STATUS_LABELS: Record<string, string> = {
  active: '活跃',
  inactive: '未激活',
  suspended: '已暂停',
  deleted: '已删除',
}

const PLATFORM_PERMISSIONS: Record<PlatformRole, string[]> = {
  super_admin: [
    'users.manage',
    'users.assign_role',
    'users.reset_password',
    'projects.manage',
    'tasks.manage',
    'export.run',
    'quality.run',
  ],
  admin: [
    'users.manage',
    'users.assign_role',
    'users.reset_password',
    'projects.manage',
    'tasks.manage',
    'export.run',
    'quality.run',
  ],
  manager: ['projects.manage', 'tasks.manage', 'export.run'],
  reviewer: ['tasks.review'],
  annotator: [],
}

export function isAdminRole(role?: string | null): boolean {
  return role === 'super_admin' || role === 'admin'
}

export function hasPermission(role: string | undefined | null, permission: string): boolean {
  if (!role) return false
  const perms = PLATFORM_PERMISSIONS[role as PlatformRole] ?? []
  return perms.includes(permission)
}

/** 当前操作者可分配的角色 */
export function assignableRoles(actorRole?: string | null): PlatformRole[] {
  if (actorRole === 'super_admin') {
    return ['super_admin', 'admin', 'manager', 'annotator', 'reviewer']
  }
  if (actorRole === 'admin') {
    return ['admin', 'manager', 'annotator', 'reviewer']
  }
  return []
}

export const MIN_PASSWORD_LENGTH = 6

export function validatePassword(password: string): string | null {
  if (password.length < MIN_PASSWORD_LENGTH) {
    return `密码至少 ${MIN_PASSWORD_LENGTH} 位`
  }
  return null
}
